meshio/vtu/_vtu.py

```diff
@@ -53,6 +54,93 @@ def _cells_from_data(connectivity, offsets, types, cell_data_raw):
         if meshio_type == "polygon" or meshio_type == "polyhedron":
             # Polygons and polyhedra have unknown and varying number of nodes per cell.
 
+            if meshio_type == "polyhedron":
+                # The data format for face-cells is:
+                # num_faces_cell_0,
+                #   num_nodes_face_0, node_ind_0, node_ind_1, ..
+                #   num_nodes_face_1, node_ind_0, node_ind_1, ..
+                #   ...
+                # num_faces_cell_1,
+                #   ...
+                #
+                # The faceoffsets describes the end of the face description for each
+                # cell, but this is 1-offset, so it is effectively the start of the
+                # next cell.
+                #
+                # See https://vtk.org/Wiki/VTK/Polyhedron_Support for more.
+
+                # There is a tacit assumption here that the cell-data keys 'faces'
+                # and 'faceoffsets' are reserved for this purpose.
+                # A .vtu file which uses 'faces' to specify the face-cell relation
+                # of a polyhedron, and also to give other type of information for
+                # other cells may lead to random behavior, at best a strange error,
+                # here.
+                # Whether such a file is permissible under the .vtu standard has not
+                # been checked, we will take our chances.
+                try:
+                    # pop removes the fields that were added to the raw cell data
+                    # in _organize_cells()
+                    faces = cell_data_raw.pop("faces")
+                    faceoffsets = cell_data_raw.pop("faceoffsets")
+                    # Switch faceoffsets to give start points, not end points
+                    faceoffsets = numpy.append(0, faceoffsets[:-1])
+                    # Take note that we have found the faces
+                    found_faces_before = True
+                except KeyError:
+                    # We will end up here in the second iteration of the loop of types.
```

Comment:
Thanks for the explanation, I wouldn't have understood it otherwise. I'd vote for refactoring this a bit. Have a  variable `is_first_iteration` or `is_second_iteration` and `if` our way around the code. Seems more explicit to me. I'm not 100% on this, what do you think?

Reply:
Your suggestion is far better than the current code. Or we can move the popping out of the main for-loop, as suggested above.

Reply:
I ended up with a version of your suggestion.